hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/DiskIOUtilManager.java

```diff
@@ -0,0 +1,253 @@
+      }
+    }
+
+    @Override
+    public String toString() {
+      return location.toString() + " disk: " + diskName;
+    }
+    @Override
+    public int hashCode() {
+      return location.hashCode();
+    }
+  }
+
+//  private String getDiskName(String path) {
+//    SystemInfo systemInfo = new SystemInfo();
+//    systemInfo.getOperatingSystem().getFileSystem().getFileStores();
+//    return "";
+//  }
+
+  private class IOStat {
+    private String diskName;
+    private long lastTotalTicks;
+    private int util;
+    public IOStat(String diskName, long lastTotalTicks) {
+      this.diskName = diskName;
+      this.lastTotalTicks = lastTotalTicks;
+    }
+
+    public int getUtil() {
+      return util;
+    }
+
+    public void setUtil(int util) {
+      if (util <= 100 && util >= 0) {
+        this.util = util;
+      } else if (util < 0) {
+        this.util = 0;
+      } else {
+        this.util = 100;
+      }
+    }
+
+    public long getLastTotalTicks() {
+      return lastTotalTicks;
+    }
+
+    public void setLastTotalTicks(long lastTotalTicks) {
+      this.lastTotalTicks = lastTotalTicks;
+    }
+  }
+
+  DiskIOUtilManager(DataNode dataNode, Configuration conf) {
+    this.dataNode = dataNode;
+    this.intervalMs = TimeUnit.SECONDS.toMillis(conf.getLong(
+        DFS_DATANODE_DISK_STAT_INTERVAL_SECONDS_KEY,
+        DFS_DATANODE_DISK_STAT_INTERVAL_SECONDS_DEFAULT));
+    if (this.intervalMs < DFS_DATANODE_DISK_STAT_INTERVAL_SECONDS_DEFAULT) {
+      this.intervalMs = 1;
+    }
+  }
+
+  @Override
+  public void run() {
+    FsVolumeImpl.LOG.info(this + " starting disk util stat");
+    while (true) {
+      if (shouldStop) {
+        FsVolumeImpl.LOG.info(this + " stopping disk util stat");
+        break;
+      }
+      if (!Shell.LINUX) {
+        FsVolumeImpl.LOG.debug("Not support disk util stat on this os release");
+        continue;
+      }
+      Map<String, IOStat> allIOStats = getDiskIoUtils();
+      synchronized (this) {
+        for (Map.Entry<DiskLocation, IOStat> entry : ioStats.entrySet()) {
+          String diskName = entry.getKey().diskName;
+          IOStat oldStat = entry.getValue();
+          int util = 0;
+          if (allIOStats.containsKey(diskName)) {
+            long oldTotalTicks = oldStat.getLastTotalTicks();
+            long newTotalTicks = allIOStats.get(diskName).getLastTotalTicks();
+            if (oldTotalTicks != 0) {
+              util = (int) ((double) (newTotalTicks - oldTotalTicks) * 100 / intervalMs);
+            }
+            oldStat.setLastTotalTicks(newTotalTicks);
+            oldStat.setUtil(util);
+            LOG.debug(diskName + " disk io util:" + util);
+          } else {
+            //Maybe this disk has been umounted.
+            oldStat.setUtil(100);
+          }
+        }
+      }
+      try {
+        Thread.sleep(intervalMs);
+      } catch (InterruptedException e) {
+      }
+    }
```

Comment:
Can you add a try..catch block to wrap this method, and log an ERROR message if any Throwable is thrown from this thread? That'll make troubleshooting easier. Otherwise, this thread could exit without any messages.